src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -131,6 +132,12 @@ impl Coordinator {
             })
             .collect::<BTreeMap<_, _>>();
 
+        // Track any external dependencies we want to provision before committing the create to the
+        // catalog. We will also ensure we retain information to deprovision them in the case of
+        // failure.
```

Comment:
Just flagging for posterity my comment about philosophy in https://github.com/MaterializeInc/materialize/issues/20019#issuecomment-1598016967. 

I find this kind of code quite scary because it will appear to work 99.9% of the time, and especially in tests. In CI, we'll successfully clean up any topics/replication slots that we create. But Murphy's law will strike in production—somehow `environmentd` will develop a habit of crashing before the `external_deprovisions` are processed, or folks will _really_ like trying to create sinks at 4pm ET right before the maintenance window, or some DBA will revoke Materialize's permissions exactly between the command to create the replication slot and the command to drop it.

Reply:
> Rather than doing this work in purification/before source creation as this sketch shows, wondering if there's something here that we might want to do on `DROP`. We could let users drop sources and free themselves from the billing burden and persistently try to clean up any state we know can be left behind.

Yeah, exactly! I posted a similar thought on #20019 and then realized @bkirwi already had that thought memorialized in a dedicated issue about this: #14551.